lib/src/simple_op_store.rs

```diff
@@ -84,281 +203,518 @@ fn not_found_to_store_error(err: std::io::Error) -> OpStoreError {
-        let temp_file = NamedTempFile::new_in(&self.path)?;
+        self.delegate.write_view(view)
+    }
+
+    fn read_operation(&self, id: &OperationId) -> OpStoreResult<Operation> {
+        self.delegate.read_operation(id)
+    }
+
+    fn write_operation(&self, operation: &Operation) -> OpStoreResult<OperationId> {
+        self.delegate.write_operation(operation)
+    }
+}
+
+#[derive(Debug)]
+struct ThriftOpStore {
+    path: PathBuf,
+}
+
+impl ThriftOpStore {
+    fn init(store_path: PathBuf) -> Self {
+        fs::create_dir(store_path.join("views")).unwrap();
+        fs::create_dir(store_path.join("operations")).unwrap();
+        Self::load(store_path)
+    }
 
-        let proto = view_to_proto(view);
-        let mut proto_bytes: Vec<u8> = Vec::new();
-        proto.write_to_writer(&mut proto_bytes)?;
+    fn load(store_path: PathBuf) -> Self {
+        ThriftOpStore { path: store_path }
+    }
 
-        temp_file.as_file().write_all(&proto_bytes)?;
+    fn view_path(&self, id: &ViewId) -> PathBuf {
+        self.path.join("views").join(id.hex())
+    }
 
-        let id = ViewId::new(Blake2b512::digest(&proto_bytes).to_vec());
+    fn operation_path(&self, id: &OperationId) -> PathBuf {
+        self.path.join("operations").join(id.hex())
+    }
+}
+
+impl OpStore for ThriftOpStore {
+    fn read_view(&self, id: &ViewId) -> OpStoreResult<View> {
+        let path = self.view_path(id);
+        let mut file = File::open(path).map_err(not_found_to_store_error)?;
+        let thrift_view = read_thrift(&mut file)?;
+        Ok(View::from(&thrift_view))
+    }
 
+    fn write_view(&self, view: &View) -> OpStoreResult<ViewId> {
+        let id = hash_view(view);
+        let temp_file = NamedTempFile::new_in(&self.path)?;
+        let thrift_view = simple_op_store_model::View::from(view);
+        write_thrift(&thrift_view, &mut temp_file.as_file())?;
         persist_content_addressed_temp_file(temp_file, self.view_path(&id))?;
         Ok(id)
     }
 
     fn read_operation(&self, id: &OperationId) -> OpStoreResult<Operation> {
         let path = self.operation_path(id);
         let mut file = File::open(path).map_err(not_found_to_store_error)?;
-
-        let proto: crate::protos::op_store::Operation = Message::parse_from_reader(&mut file)?;
-        Ok(operation_from_proto(&proto))
+        let thrift_operation = read_thrift(&mut file)?;
+        Ok(Operation::from(&thrift_operation))
     }
 
     fn write_operation(&self, operation: &Operation) -> OpStoreResult<OperationId> {
+        let id = hash_operation(operation);
         let temp_file = NamedTempFile::new_in(&self.path)?;
+        let thrift_operation = simple_op_store_model::Operation::from(operation);
+        write_thrift(&thrift_operation, &mut temp_file.as_file())?;
+        persist_content_addressed_temp_file(temp_file, self.operation_path(&id))?;
+        Ok(id)
+    }
+}
+
+fn hash_ref_target(ref_target: &RefTarget) -> Vec<u8> {
```

Comment:
Should these hashers mostly be methods on the types in question? Especially if they might be used by other op store impls.

Reply:
I'd prefer to wait until we have another implementation (most likely Google's internal one). It might be confusing if they're on the object if one implementation uses a different hashing scheme (or maybe not even hash-based identifiers).